### PR TITLE
[Application QuickStart] [Backend] DAB Connection string transformation for Docker containers

### PR DESCRIPTION
**Pull request by manujoseph85**
## Description

The default connection string from the Object Explorer will point to localhost (or other forms) for local sqlserver deployment. For DAB deployed in Docker we need to transform this for DAB (to host.docker.internal) to be able to redirect to host correctly.

* Added a new `sqlServerContainerName` property to the `Dab.DabConnectionInfo` interface
* transformConnectionInfoForDocker in dabService now checks for localhost and its variants in the provided connection string and transforms it to
  * host.docker.internal\\${sqlServerContainerName} for sqlserver in local containers
  * host.docker.internal for local sqlserver instance

**Testing and validation:**

* Added extensive unit tests for connection string transformation logic, covering edge cases, localhost variants, containerized SQL Server, and integration with config generation.

## Code Changes Checklist

-   [x] New or updated **unit tests** added
-   [x] All existing tests pass (`npm run test`)
-   [x] Code follows [contributing guidelines](https://github.com/microsoft/vscode-mssql/blob/main/CONTRIBUTING.md)
-   [x] Telemetry/logging updated if relevant
-   [x] No regressions or UX breakage

### CODE REVIEW

File: extensions/mssql/test/unit/dab/dabService.test.ts [hunk 1]
```diff
@@ -77,6 +77,230 @@ suite("DabService Tests", () => {
...
+
+            test("should not add container name when server is not localhost", () => {
+                const result = transform(
+                    "Server=remotehost.example.com;Database=TestDb;",
+                    "my-sql-container",
+                );
+                expect(result.connectionString).to.include("remotehost.example.com");
+                expect(result.connectionString).to.not.include("host.docker.internal");
+            });
+        });
```
[the remark this answers is not in the document]
Comment (manujoseph85) [reply extensions/mssql/test/unit/dab/dabService.test.ts:253]: Duplicate comment

File: extensions/mssql/src/services/dabService.ts [hunk 2]
```diff
@@ -305,4 +312,97 @@ export class DabService implements Dab.IDabService {
...
+        const host = this.parseHostFromServerValue(serverValue);
+
+        // Check if this is a localhost address
+        if (!this.isLocalhostAddress(host)) {
+            return connectionInfo;
+        }
+
+        // Always use host.docker.internal to reach services on the host machine.
+        // For containerized SQL Server, append the container name as a suffix
+        // since port mapping exposes it on the host.
```
[the remark this answers is not in the document]
Comment (manujoseph85) [reply extensions/mssql/src/services/dabService.ts]: Done

File: extensions/mssql/test/unit/schemaDesignerWebviewController.test.ts [hunk 3]
```diff
@@ -686,14 +686,65 @@ suite("SchemaDesignerWebviewController tests", () => {
...
+                    treeNode,
+                    connectionUri,
+                );
+
+                const handler = requestHandlers.get(Dab.GenerateConfigRequest.type.method);
+                const result = await handler({ config: mockDabConfig });
+
+                const parsedConfig = JSON.parse(result.configContent);
+                expect(parsedConfig["data-source"]["connection-string"]).to.equal(
+                    remoteConnectionString,
```
[the remark this answers is not in the document]
Comment (manujoseph85) [reply extensions/mssql/test/unit/schemaDesignerWebviewController.test.ts:746]: Fixed

File: extensions/mssql/src/services/dabService.ts [hunk 4]
```diff
@@ -305,4 +312,97 @@ export class DabService implements Dab.IDabService {
...
+        const newHost = sqlServerContainerName
+            ? `host.docker.internal\\${sqlServerContainerName}`
+            : "host.docker.internal";
```
[the remark this answers is not in the document]
Comment (manujoseph85) [reply extensions/mssql/src/services/dabService.ts]: Done

File: extensions/mssql/src/services/dabService.ts [hunk 5]
```diff
@@ -305,4 +312,97 @@ export class DabService implements Dab.IDabService {
...
+
+        // Replace the host portion in the server value, preserving port and instance name
+        const newServerValue = serverValue.replace(
+            new RegExp(`^${this.escapeRegex(host)}`, "i"),
+            newHost,
+        );
+
```
[the remark this answers is not in the document]
Comment (manujoseph85) [reply extensions/mssql/src/services/dabService.ts]: Fixed

**Review by caohai — DISMISSED** [review]

File: extensions/mssql/src/services/dabService.ts [hunk 6]
```diff
@@ -305,4 +313,104 @@ export class DabService implements Dab.IDabService {
...
+        // Check if this is a localhost address
+        if (!this.isLocalhostAddress(host)) {
+            return connectionInfo;
+        }
+
+        // Build the new server value:
+        // - Always use host.docker.internal to reach services on the host machine.
+        // - For containerized SQL Server, replace host and any instance name with
+        //   host.docker.internal\<container-name>, preserving only the port.
+        //   Containerized SQL Server does not use named instances.
```
[the remark this answers is not in the document]
Comment (manujoseph85) [reply extensions/mssql/src/services/dabService.ts:349]: Will do separate testing on Linux

File: extensions/mssql/src/services/dabService.ts [hunk 7]
```diff
@@ -57,18 +65,18 @@ export class DabService implements Dab.IDabService {
...
         params?: Dab.DabDeploymentParams,
         config?: Dab.DabConfig,
-        connectionString?: string,
+        connectionInfo?: Dab.DabConnectionInfo,
     ): Promise<Dab.RunDeploymentStepResponse> {
         // Generate config content if needed for startContainer step
         let configContent: string | undefined;
-        if (step === Dab.DabDeploymentStepOrder.startContainer && config && connectionString) {
-            const configResponse = this.generateConfig(config, { connectionString });
+        if (step === Dab.DabDeploymentStepOrder.startContainer && config && connectionInfo) {
```
[the remark this answers is not in the document]
Comment (manujoseph85) [reply extensions/mssql/src/services/dabService.ts:78]: Testing connectionInfo is sufficient

File: extensions/mssql/src/services/dabService.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (manujoseph85) [reply extensions/mssql/src/services/dabService.ts:349]: Added explicit HostConfig for Linux

File: extensions/mssql/src/services/dabService.ts [hunk 8]
```diff
@@ -305,4 +313,104 @@ export class DabService implements Dab.IDabService {
...
+        // Supports both "Server=" and "Data Source=" formats
+        const serverMatch = connectionString.match(/(?:Server|Data Source)\s*=\s*([^;]+)/i);
+        if (!serverMatch) {
+            return connectionInfo;
+        }
+
+        const serverValue = serverMatch[1].trim();
+
+        // Parse the server address to check if it's localhost
+        const host = this.parseHostFromServerValue(serverValue);
```
Comment (aasimkhan30) [inline extensions/mssql/src/services/dabService.ts:338]: Isn't this information already in connectionInfo?

File: extensions/mssql/src/services/dabService.ts [hunk 9]
```diff
@@ -305,4 +313,104 @@ export class DabService implements Dab.IDabService {
...
+        // Build the new server value:
+        // - Always use host.docker.internal to reach services on the host machine.
+        // - For containerized SQL Server, replace host and any instance name with
+        //   host.docker.internal\<container-name>, preserving only the port.
+        //   Containerized SQL Server does not use named instances.
+        // - For host SQL Server, replace only the host, preserving instance name and port.
+        let newServerValue: string;
+        if (sqlServerContainerName) {
+            // Extract port if present (comma-separated), discard any instance name
+            const commaIndex = serverValue.indexOf(",");
```
Comment (aasimkhan30) [inline extensions/mssql/src/services/dabService.ts:354]: Please use the connection info object from to get these details directly instead of parsing it from connection string. 

I think we should use the original connection info, clone it, mutate it and then create a connection string out of it. It will be less brittle than regexes. 

https://github.com/microsoft/vscode-mssql/blob/a29a77fde34bb8943ec5ceb52c1ea6f7f6584dda/extensions/mssql/src/connectionconfig/connectionDialogWebviewController.ts#L606-L627

**Review by aasimkhan30 — COMMENTED** [review]
Looks good with minor comments.

File: extensions/mssql/src/services/dabService.ts (hunk 9 above)
Comment (manujoseph85) [reply extensions/mssql/src/services/dabService.ts:354]: Yes, I think thats probably the best way to do it too.

Although, we might have to do some more plumbing in SchemaDesignerWebviewController to do that whereas connectionString was already available. We are only replacing in a narrow case - Server part when its localhost. So, I think the regex matches here are less error prone.

So, I will merge this for now and file an improvement item for using connectionManager to do the rewrite.

File: extensions/mssql/src/services/dabService.ts (hunk 8 above)
Comment (manujoseph85) [reply extensions/mssql/src/services/dabService.ts:338]: connectionString was already available in DabService. I will file an improvement item for using connectionManager to do the rewrite.
